File: libs/power-log-parser/src/lib/handlers/action-metadata-info-handler.ts
import type { Helper } from '../helper';
import { Info, MetaData, Node } from '../models';
import { Regexes } from '../regexes';
import type { ParserState } from '../state/parser-state';

export class ActionMetadataInfoHandler {
	static Handle(
		timestamp: string,
		data: string,
		state: ParserState,
		indentLevel: number,
		helper: Helper,
	): boolean {
		const match = Regexes.ActionMetaDataInfoRegex.exec(data);
		if (match) {
			const index = match[1];
			const rawEntity = match[2];
			const entity = helper.ParseEntity(rawEntity);
			const metaInfo = new Info();
			metaInfo.Id = entity;
			metaInfo.Index = parseInt(index, 10);
			metaInfo.Entity = entity;
			metaInfo.TimeStamp = timestamp;

			if (state.Node!.Type === MetaData) {
				(state.Node!.Object as MetaData).MetaInfo.push(metaInfo);
			} else {
				throw new Error('Invalid node ' + state.Node!.Type + ' while parsing ' + data);
			}
			state.CreateNewNode(new Node(Info, metaInfo, indentLevel, state.Node, data));
			return true;
		}
		return false;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Parser/Handlers/DataHandler/ActionMetadataInfoHandler.cs
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser.ReplayData;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;
using HearthstoneReplays.Parser.ReplayData.Meta;

namespace HearthstoneReplays.Parser.Handlers
{
    internal class ActionMetadataInfoHandler
    {
        public static bool HandleActionMetaDataInfo(DateTime timestamp, string data, ParserState state, int indentLevel, Helper helper)
        {
            var match = Regexes.ActionMetaDataInfoRegex.Match(data);
            if (match.Success)
            {
                var index = match.Groups[1].Value;
                var rawEntity = match.Groups[2].Value;
                var entity = helper.ParseEntity(rawEntity);
                var metaInfo = new Info { Id = entity, Index = int.Parse(index), Entity = entity, TimeStamp = timestamp };
                if (state.Node.Type == typeof(MetaData))
                    ((MetaData)state.Node.Object).MetaInfo.Add(metaInfo);
                else
                    throw new Exception("Invalid node " + state.Node.Type + " while parsing " + data);
                state.CreateNewNode(new Node(typeof(Info), metaInfo, indentLevel, state.Node, data));
                return true;
            }
            return false;
        }
    }
}
